llvm/lib/Target/BPF/BPFInstrInfo.td

```diff
@@ -522,6 +526,28 @@ let Predicates = [BPFNoALU32] in {
 }
 def STD : STOREi64<BPF_DW, "u64", store>;
 
+class relaxed_store<PatFrag base>
+  : PatFrag<(ops node:$val, node:$ptr), (base node:$val, node:$ptr)> {
+  let IsAtomic = 1;
+  let IsAtomicOrderingReleaseOrStronger = 0;
+}
+
+class releasing_store<PatFrag base>
+  : PatFrag<(ops node:$val, node:$ptr), (base node:$val, node:$ptr)> {
+  let IsAtomic = 1;
+  let IsAtomicOrderingRelease = 1;
+}
```

Comment:
I found the following in arm64:
```
// A store operation that actually needs release semantics.
class releasing_store<PatFrag base>
  : PatFrag<(ops node:$ptr, node:$val), (base node:$val, node:$ptr)> {
  let IsAtomic = 1;
  let IsAtomicOrderingReleaseOrStronger = 1;
}

// An atomic store operation that doesn't actually need to be atomic on AArch64.
class relaxed_store<PatFrag base>
  : PatFrag<(ops node:$ptr, node:$val), (base node:$val, node:$ptr)> {
  let IsAtomic = 1;
  let IsAtomicOrderingReleaseOrStronger = 0;
}
```

What is the difference between IsAtomicOrderingReleaseOrStronger = 1 vs. IsAtomicOrderingReleas = 1?

Reply:
For `ATOMIC_STORE`: `IsAtomicOrderingRelease = 1` only matches `RELEASE`. 
 `IsAtomicOrderingReleaseOrStronger = 1` matches both `RELEASE` and `SEQ_CST` (sequentially consistent).
- - -
For ARM64, it looks like that `releasing_store<>` has been intended to match both `RELEASE` and `SEQ_CST` since at least commit 00ed9964c659 ("ARM64: initial backend import"):

```c
// A store operation that actually needs release semantics.
class releasing_store<PatFrag base>
  : PatFrag<(ops node:$ptr, node:$val), (base node:$ptr, node:$val), [{
  AtomicOrdering Ordering = cast<AtomicSDNode>(N)->getOrdering();
  assert(Ordering != AcquireRelease && "unexpected store ordering");
  return Ordering == Release || Ordering == SequentiallyConsistent;
}]>;
```
Looking at `llvm/test/CodeGen/AArch64/fast-isel-atomic.ll`, it seems that a "`store atomic` `release`" and a "`store atomic` `seq_cst`" can be turned into the same ARM64 insn:
```
; CHECK-LABEL: atomic_store_release_16:
; CHECK-NEXT: // %bb.0:
; CHECK-NEXT:  stlrh w1, [x0]
; CHECK-NEXT:  ret
define void @atomic_store_release_16(ptr %p, i16 %val) #0 {
  store atomic i16 %val, ptr %p release, align 2
  ret void
}
```
```
; CHECK-LABEL: atomic_store_seq_cst_16:
; CHECK-NEXT: // %bb.0:
; CHECK-NEXT:  stlrh w1, [x0]
; CHECK-NEXT:  ret
define void @atomic_store_seq_cst_16(ptr %p, i16 %val) #0 {
  store atomic i16 %val, ptr %p seq_cst, align 2
  ret void
}
```
I haven't read the manual to understand why yet, however.
- - -
For us, I used `IsAtomicOrderingRelease` since I didn't want to match `seq_cst` (the 4th commit in this PR makes it generate an "unsupported" error if user called `__atomic_{load,store}[_n]()` with `__ATOMIC_SEQ_CST`, as suggested by Eduard).

Reply:
Okay, I see. It is interesting that arm64 generates the same insn with release and seq_cst. In my thinking, we should just support __ATOMIC_ACQUIRE and __ATOMIC_RELEASE since currently with __ATOMIC_RELAXED, e.g. __atomic_load_n(ptr, __ATOMIC_RELAXED), the actually generated code should be '*ptr'. Is my interpretation correct or  __atomic_load_n(ptr, __ATOMIC_RELAXED) could have semantics beyond '*ptr'?

Reply:
Eduard [suggested earlier](https://github.com/llvm/llvm-project/pull/108636#discussion_r1765597286) (referencing [Andrii's suggestion](https://github.com/llvm/llvm-project/pull/107343#issuecomment-2334664935) in #107343) that, for now, we "fall back" to `ACQUIRE` or `RELEASE` if user requested `RELAXED`, until we actually support `RELAXED` in the future.

> Is my interpretation correct or __atomic_load_n(ptr, __ATOMIC_RELAXED) could have semantics beyond '*ptr'?

`__ATOMIC_RELAXED` still guarantees atomicity but `*ptr` does not, if i'm understanding correctly?

Reply:
> Eduard [suggested earlier](https://github.com/llvm/llvm-project/pull/108636#discussion_r1765597286) (referencing [Andrii's suggestion](https://github.com/llvm/llvm-project/pull/107343#issuecomment-2334664935) in #107343) that, for now, we "fall back" to `ACQUIRE` or `RELEASE` if user requested `RELAXED`, until we actually support `RELAXED` in the future.
> 
> > Is my interpretation correct or __atomic_load_n(ptr, __ATOMIC_RELAXED) could have semantics beyond '*ptr'?
> 
> `__ATOMIC_RELAXED` still guarantees atomicity but `*ptr` does not, if i'm understanding correctly?

bpf prog does not support misaligned load/store etc. So for 'int *ptr', 'long *ptr', etc. dereference will have a single load or store and it is atomic.

Reply:
Got it, thanks for the input!  I also found this in [Instruction-Level BPF Memory Model](https://docs.google.com/document/d/1TaSEfWfLnRUi5KqkavUQyL2tThJXYWHS15qcbxIsFb0/edit?usp=sharing), section "Atomic Loads" :

> For `__ATOMIC_RELAXED`, these can simply be implemented as plain BPF load instructions.